rust/rbac-registration/src/cardano/cip509/cip509.rs

```diff
@@ -0,0 +1,572 @@
+
+use std::{borrow::Cow, collections::HashMap};
+
+use anyhow::{anyhow, Context};
+use cardano_blockchain_types::{MetadatumLabel, MultiEraBlock, TxnIndex};
+use catalyst_types::{
+    hashes::{Blake2b256Hash, BLAKE_2B256_SIZE},
+    problem_report::ProblemReport,
+};
+use cbork_utils::decode_helper::{decode_bytes, decode_helper, decode_map_len};
+use minicbor::{
+    decode::{self},
+    Decode, Decoder,
+};
+use pallas::{
+    codec::utils::Nullable,
+    ledger::{
+        addresses::{Address, ShelleyAddress},
+        primitives::conway,
+        traverse::MultiEraTx,
+    },
+};
+use strum_macros::FromRepr;
+use tracing::warn;
+use uuid::Uuid;
+
+use crate::{
+    cardano::cip509::{
+        decode_context::DecodeContext,
+        rbac::Cip509RbacMetadata,
+        types::{PaymentHistory, RoleNumber, TxInputHash, ValidationSignature},
+        utils::Cip0134UriSet,
+        validation::{
+            validate_aux, validate_role_data, validate_stake_public_key, validate_txn_inputs_hash,
+        },
+        x509_chunks::X509Chunks,
+        Payment, PointTxnIdx, RoleData,
+    },
+    utils::decode_helper::{report_duplicated_key, report_missing_keys},
+};
+
+/// A x509 metadata envelope.
+///
+/// The envelope is required to prevent replayability attacks. See [this document] for
+/// more details.
+///
+/// [this document]: https://github.com/input-output-hk/catalyst-CIPs/blob/x509-envelope-metadata/CIP-XXXX/README.md
+#[derive(Debug, Clone)]
+#[allow(clippy::module_name_repetitions)]
+pub struct Cip509 {
+    /// A registration purpose (`UUIDv4`).
+    ///
+    /// The purpose is defined by the consuming dApp.
+    purpose: Option<Uuid>,
+    /// Transaction inputs hash.
+    txn_inputs_hash: Option<TxInputHash>,
+    /// An optional hash of the previous transaction.
+    ///
+    /// The hash must always be present except for the first registration transaction.
+    prv_tx_id: Option<Blake2b256Hash>,
+    /// Metadata.
+    ///
+    /// This field encoded in chunks. See [`X509Chunks`] for more details.
+    metadata: Option<Cip509RbacMetadata>,
+    /// A validation signature.
+    validation_signature: Option<ValidationSignature>,
+    /// A payment history.
+    ///
+    /// The history is only tracked for the addresses that are passed to `Cip509`
+    /// constructors.
+    payment_history: PaymentHistory,
+    /// A hash of the transaction from which this registration is extracted.
+    txn_hash: Blake2b256Hash,
+    /// A point (slot) and a transaction index identifying the block and the transaction
+    /// that this `Cip509` was extracted from.
+    origin: PointTxnIdx,
+    /// A report potentially containing all the issues occurred during `Cip509` decoding
+    /// and validation.
+    ///
+    /// The data located in `Cip509` is only considered valid if
+    /// `ProblemReport::is_problematic()` returns false.
+    report: ProblemReport,
```

Comment:
Seems redundant at this place as we are storing it inside the `DecodeContext` struct.
Also by storing problem report at this place we assuming that it could be a `Cip509` invalidly structured object, what I think its better to avoid and keep with the following logic.
- trying to decode bytes to `Cip509`.
- if it fails getting an error and problem report, not getting a `Cip509` object.
- if it is ok, getting a valid (at least structurally) `Cip509`.